File: src/types/action.ts
export type ConflictResolution = "rename" | "replace" | "skip";

export interface MoveAction {
  destination: string;
  onConflict: ConflictResolution;
  skipDuplicates: boolean;
}

export interface CopyAction {
  destination: string;
  onConflict: ConflictResolution;
  skipDuplicates: boolean;
}

export interface RenameAction {
  pattern: string;
  onConflict: ConflictResolution;
}

export interface SortAction {
  destination: string;
  onConflict: ConflictResolution;
}

export type ArchiveFormat = "zip" | "tar" | "tarGz";

export interface ArchiveAction {
  destination: string;
  format: ArchiveFormat;
  deleteAfter: boolean;
}

export interface UnarchiveAction {
  destination?: string;
  deleteAfter: boolean;
}

export interface DeleteAction {
  permanent: boolean;
}

export interface ScriptAction {
  command: string;
}

export interface NotifyAction {
  message: string;
}

export type OpenAction = {};

export interface PauseAction {
  durationSeconds: number;
}

export type Action =
  | { type: "move" } & MoveAction
  | { type: "copy" } & CopyAction
  | { type: "rename" } & RenameAction
  | { type: "sortIntoSubfolder" } & SortAction
  | { type: "archive" } & ArchiveAction
  | { type: "unarchive" } & UnarchiveAction
  | { type: "delete" } & DeleteAction
  | { type: "deletePermanently" } & DeleteAction
  | { type: "runScript" } & ScriptAction
  | { type: "notify" } & NotifyAction
  | { type: "open" } & OpenAction
  | { type: "pause" } & PauseAction
  | { type: "continue" }
  | { type: "ignore" };

export interface ActionDetails {
  sourcePath: string;
  destinationPath?: string;
  metadata: Record<string, string>;
}

export type ActionType =
  | "move"
  | "copy"
  | "rename"
  | "sortIntoSubfolder"
  | "archive"
  | "unarchive"
  | "delete"
  | "deletePermanently"
  | "runScript"
  | "notify"
  | "open"
  | "pause"
  | "continue"
  | "ignore";

File: src-tauri/src/models/action.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "camelCase")]
pub enum Action {
    Move(MoveAction),
    Copy(CopyAction),
    Rename(RenameAction),
    SortIntoSubfolder(SortAction),
    Archive(ArchiveAction),
    Unarchive(UnarchiveAction),
    Delete(DeleteAction),
    DeletePermanently(DeleteAction),
    RunScript(ScriptAction),
    Notify(NotifyAction),
    Open(OpenAction),
    Pause(PauseAction),
    Continue,
    Ignore,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MoveAction {
    pub destination: String,
    pub on_conflict: ConflictResolution,
    pub skip_duplicates: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CopyAction {
    pub destination: String,
    pub on_conflict: ConflictResolution,
    pub skip_duplicates: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RenameAction {
    pub pattern: String,
    pub on_conflict: ConflictResolution,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SortAction {
    pub destination: String,
    pub on_conflict: ConflictResolution,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ArchiveAction {
    pub destination: String,
    pub format: ArchiveFormat,
    pub delete_after: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UnarchiveAction {
    pub destination: Option<String>,
    pub delete_after: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DeleteAction {
    pub permanent: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ScriptAction {
    pub command: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NotifyAction {
    pub message: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct OpenAction {}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PauseAction {
    pub duration_seconds: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub enum ArchiveFormat {
    Zip,
    Tar,
    #[serde(rename = "tarGz")]
    TarGz,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum ConflictResolution {
    Rename,
    Replace,
    Skip,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ActionDetails {
    pub source_path: String,
    pub destination_path: Option<String>,
    pub metadata: HashMap<String, String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub enum ActionType {
    Move,
    Copy,
    Rename,
    SortIntoSubfolder,
    Archive,
    Unarchive,
    Delete,
    DeletePermanently,
    RunScript,
    Notify,
    Open,
    Pause,
    Continue,
    Ignore,
}
